devito/ir/iet/nodes.py

```diff
@@ -273,51 +261,55 @@ def __repr__(self):
         return "%sCall::\n\t%s(...)" % (ret, self.name)
 
     @property
+    def children(self):
+        return tuple(i for i in self.arguments if isinstance(i, (Call, Lambda)))
+
+    @cached_property
     def functions(self):
         retval = []
         for i in self.arguments:
-            if isinstance(i, numbers.Number):
-                continue
-            elif isinstance(i, (AbstractFunction, Indexed, LocalObject)):
+            if isinstance(i, (AbstractFunction, Indexed, LocalObject)):
                 retval.append(i.function)
+            elif isinstance(i, Call):
+                retval.extend(i.functions)
             else:
-                for s in i.free_symbols:
+                try:
+                    v = i.free_symbols
+                except AttributeError:
+                    continue
+                for s in v:
                     try:
-                        f = s.function
+                        # `try-except` necessary for e.g. Macro
+                        if isinstance(s.function, AbstractFunction):
+                            retval.append(s.function)
```

Comment:
then why not directly
```
try:
    retval.append(s.function)
except AttributeError
```

Reply:
because if it's not an AbstractFunction we don't want to collect the `.function`. The `.function` property is provided by all our `types.Basic` objects. One day we could purge some of this, but for now I need to pick the _actual_ Functions only here

Reply:
> One day we could purge some of this

I mean, the `types.Basic` hierarchy, by e.g. removing the .function from some types, but it's less trivial than it sounds I'm afraid